### PR TITLE
Implement hwinfo

### PR DESCRIPTION
**Pull request by djoreilly**
This will be used on x86_64 and arm64 to find the CPU and socket
count, and "Hypervisor vendor" on VMs. A map[string]string is
returned. The callers can lookup relevant keys and convert the
value to another type as needed.

### CODE REVIEW

**Comment by djoreilly** [comment]
There are still more commits to come on this branch.

File: connect/hwinfo.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
+package connect
+
+import (
+	"bufio"
+	"bytes"
+	"strings"
+)
+
+func lscpu() (map[string]string, error) {
+	output, err := execute([]string{"lscpu"}, false, nil)
```
Comment (skazi0) [inline connect/hwinfo.go:26]: `lscpu` has localized outputs. We should probably set some env variables to make sure we get expected output for parsing.

File: connect/hwinfo.go [hunk 2]
```diff
@@ -88,3 +89,42 @@ func uuid() (string, error) {
 	}
 	return string(output), nil
 }
+
+// getPrivateIPAddr returns the first private IP address on the host
+func getPrivateIPAddr() (string, error) {
+	output, err := execute([]string{"ip", "-4", "--oneline", "--brief", "address"}, false, nil)
```
Comment (skazi0) [inline connect/hwinfo.go:95]: Why not use the built in `net.InterfaceAddrs()` like the ruby version did with `Socket.ip_address_list` ?

Comment (djoreilly) [reply connect/hwinfo.go:95]: > Why not use the built in `net.InterfaceAddrs()` like the ruby version did with `Socket.ip_address_list` ?

sure - that  way looks simpler. I will change it

File: connect/hwinfo.go [hunk 3]
```diff
@@ -130,3 +130,26 @@ func hostname() string {
 	}
 	return ip
 }
+
+func readValues(arg string) (map[string]string, error) {
+	output, err := execute([]string{"read_values", arg}, false, nil)
```
Comment (skazi0) [inline connect/hwinfo.go:135]: It would be good to add some info to the code and/or commit message about this being specific to S390.

File: connect/hwinfo.go [hunk 4]
```diff
@@ -28,3 +106,114 @@ func lscpu2map(b []byte) map[string]string {
...
+	Debug.Print(err)
+	ip, err := getPrivateIPAddr()
+	if err != nil {
+		Debug.Print(err)
+		return ""
+	}
+	return ip
+}
+
+// readValues calls read_values from SUSUE/s390-tools
```
Comment (skazi0) [inline connect/hwinfo.go:196]: typo

Comment (djoreilly) [reply connect/hwinfo.go:196]: well spotted. Will fix with next push

File: connect/hwinfo.go [hunk 5]
```diff
@@ -53,3 +62,29 @@ func findCloudProvider(b []byte) string {
...
+		content, err := os.ReadFile("/sys/hypervisor/uuid")
+		if err != nil {
+			return "", err
+		}
+		return string(content), nil
+	}
+	output, err := execute([]string{"dmidecode", "-s", "system-uuid"}, false, nil)
+	if err != nil {
+		return "", err
+	}
```
Comment (djoreilly) [inline connect/hwinfo.go:151]: TODO need to check for ['Not Settable', 'Not Present']
https://github.com/SUSE/connect/blob/master/lib/suse/connect/hwinfo/x86.rb#L34

File: connect/api.go [hunk 6]
```diff
@@ -57,3 +58,46 @@ func GetProduct(productQuery Product) (Product, error) {
 	}
 	return remoteProduct, nil
 }
+
+// updateSystemAPI resend the system's hardware info to SCC.
+// https://github.com/SUSE/connect/wiki/SCC-API-%28Implemented%29#update-system
+func updateSystemAPI(distroTarget, namespace string, instanceData []byte) error {
```
Comment (djoreilly) [inline connect/api.go:64]: Maybe putSystem() is better here.

Comment (skazi0) [reply connect/api.go:64]: why not just `updateSystem()` ?

Comment (skazi0) [reply connect/api.go:64]: btw, the link in the comment doesn't seem to be valid anymore

File: connect/api.go [hunk 7]
```diff
@@ -57,3 +58,46 @@ func GetProduct(productQuery Product) (Product, error) {
...
+	if err != nil {
+		return err
+	}
+	body := bytes.NewReader(payload)
+	_, err = callHTTP("PUT", "/connect/systems", body, nil, authSystem)
+	return err
+}
+
+// updatePayload returns the JSON body needed for the announce and update calls
+func updatePayload(distroTarget, namespace string, instanceData []byte) ([]byte, error) {
```
Comment (skazi0) [inline connect/api.go:75]: this should be something like `systemUpdatePayload()` as there are other update APIs like `PUT /connect/systems/products`

File: connect/api.go (hunk 6 above)
Comment (djoreilly) [reply connect/api.go:64]: > why not just `updateSystem()` ?

because I have already taken that name in client.go, so it needs to something different. I like putSystem()  here, and it follows a convention we already have with GetProduct() and GetActivations()

Comment (djoreilly) [reply connect/api.go:64]: > > why not just `updateSystem()` ?
> 
> because I have already taken that name in client.go, so it needs to something different. I like putSystem() here, and it follows a convention we already have with GetProduct() and GetActivations()

thanks - I should have check that

File: connect/api.go [hunk 8]
```diff
@@ -57,3 +58,46 @@ func GetProduct(productQuery Product) (Product, error) {
...
 }
+
+// updateSystemAPI resend the system's hardware info to SCC.
+// https://github.com/SUSE/connect/wiki/SCC-API-%28Implemented%29#update-system
+func updateSystemAPI(distroTarget, namespace string, instanceData []byte) error {
+	payload, err := updatePayload(distroTarget, namespace, instanceData)
+	if err != nil {
+		return err
+	}
+	body := bytes.NewReader(payload)
```
Comment (djoreilly) [inline connect/api.go:69]: I think it would be more natural and convenient for callHTTP() to receive a []byte instead of io.Reader for the body. I will propose a change for that

File: connect/hwinfo.go (hunk 5 above)
Comment (djoreilly) [reply connect/hwinfo.go:151]: done

File: connect/api.go (hunk 8 above)
Comment (djoreilly) [reply connect/api.go]: done

File: connect/api.go (hunk 6 above)
Comment (djoreilly) [reply connect/api.go]: I made some changes around here. Maybe better?

File: connect/connection.go [hunk 9]
```diff
@@ -58,14 +59,14 @@ func successCode(code int) bool {
 	return code >= 200 && code < 300
 }
 
-func callHTTP(verb, path string, body io.Reader, query map[string]string, auth authType) ([]byte, error) {
+func callHTTP(verb, path string, body []byte, query map[string]string, auth authType) ([]byte, error) {
```
Comment (skazi0) [inline connect/connection.go:62]: Please take this change out to a separate PR to make parallel work easier and limit future rebasing.

File: connect/client.go [hunk 10]
```diff
@@ -24,3 +63,25 @@ func IsRegistered() bool {
...
+		fmt.Println("Registering system to ", server)
+	} else {
+		fmt.Println("Deregistering system from ", server)
+	}
+	if CFG.FsRoot != "" {
+		fmt.Println("Rooted at: ", CFG.FsRoot)
+	}
+
+	// TODO
+	// if CFG.Email != "" {
```
Comment (skazi0) [inline connect/client.go:84]: any idea where this email comes from?

File: connect/client.go [hunk 11]
```diff
@@ -24,3 +63,25 @@ func IsRegistered() bool {
 func URLDefault() bool {
 	return CFG.BaseURL == defaultBaseURL
 }
+
+func printInformation(action string) {
```
Comment (skazi0) [inline connect/client.go:67]: This would also be a good candidate for a separate PR. I could use it directly in my Deregister code.

File: connect/api.go (hunk 6 above)
Comment (skazi0) [reply connect/api.go]: looks good

File: connect/api.go (hunk 8 above)
Comment (skazi0) [reply connect/api.go]: nice. thx!

File: connect/connection.go (hunk 9 above)
Comment (djoreilly) [reply connect/connection.go:62]: will do

File: connect/client.go (hunk 10 above)
Comment (djoreilly) [reply connect/client.go:84]: need to research that - the cli I think

File: connect/client.go (hunk 11 above)
Comment (djoreilly) [reply connect/client.go:67]: I will split it out

File: connect/connection.go (hunk 9 above)
Comment (djoreilly) [reply connect/connection.go:62]: done

File: connect/client.go (hunk 11 above)
Comment (djoreilly) [reply connect/client.go:67]: done

File: connect/client.go (hunk 10 above)
Comment (skazi0) [reply connect/client.go]: yeah, now I see that. I need to get my vision checked :o)

**Comment by djoreilly** [comment]
rebased

**Comment by djoreilly** [comment]
Need to rebase this and make changes for the new naming convention

File: connect/api.go [hunk 12]
```diff
@@ -84,3 +84,41 @@ func deactivateProduct(product Product) (Service, error) {
 	err = json.Unmarshal(resp, &remoteService)
 	return remoteService, err
 }
+
+// updateSystem updates the system's hardware info on SCC
+// https://scc.suse.com/connect/v4/documentation#/systems/put_systems
```
Comment (skazi0) [inline connect/api.go:89]: please add a similar comment about the format of this data and `makeSysInfoBody()`

**Review by skazi0 — APPROVED** [review]
looks good so far. great work!

**Comment by djoreilly** [comment]
rebased

File: connect/api.go (hunk 12 above)
Comment (djoreilly) [reply connect/api.go:112]: done

File: connect/hwinfo.go [hunk 13]
```diff
@@ -0,0 +1,262 @@
...
+	Hypervisor    string `json:"hypervisor"`
+	Arch          string `json:"arch"`
+	UUID          string `json:"uuid"`
+	CloudProvider string `json:"cloud_provider"`
+}
+
+func getHwinfo() (hwinfo, error) {
+	hw := hwinfo{}
+	var err error
+	hw.Arch, err = arch()
```
Comment (skazi0) [inline connect/hwinfo.go:37]: with `err` defined above, these could be folded into the `if`s below.

Comment (skazi0) [reply connect/hwinfo.go:37]: i.e. 
```
if hw.Arch, err = arch(); err != nil {
    return hwinfo{}, err
}

File: connect/hwinfo.go [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+	"bytes"
+	"net"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+var (
+	cloudEx = `(?m)Version: .*(amazon)|Manufacturer: (Amazon)|Manufacturer: (Google)|Manufacturer: (Microsoft) Corporation`
```
Comment (skazi0) [inline connect/hwinfo.go:14]: I don't think it should be a multiline match.

**Review by skazi0 — APPROVED** [review]
lgtm. just some nit picking. feel free to ignore.

File: connect/hwinfo.go (hunk 14 above)
Comment (djoreilly) [reply connect/hwinfo.go:14]: yeah, I'll take it out. I can't remember what I was trying

Comment (djoreilly) [reply connect/hwinfo.go]: done

File: connect/hwinfo.go (hunk 13 above)
Comment (djoreilly) [reply connect/hwinfo.go]: done

**Review by skazi0 — APPROVED** [review]
